### PR TITLE
[MIG][16.0] product_country_restriction

### PR DESCRIPTION
**Pull request by xavier-bouquiaux**
Migration from 10

### CODE REVIEW

File: product_country_restriction/models/country_restriction_mixin.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        res = self._update_country_strategy(res)
+        return res
+
+    def _update_country_strategy(self, result):
+        """
+        This is used to apply another strategy to the rules result
+        depending on company strategy
+        :return:
+        """
+        strategy = self.env.user.company_id.country_restriction_strategy
```
Comment (acsonefho) [inline product_country_restriction/models/country_restriction_mixin.py]: IMO it's not correct: it should be `self.env.company.country_restriction_strategy`. And maybe check before if the current record has a `company_id` (and use it in priority).
Something like
```
company = self.env.company
if hasattr(self, 'company_id') and self.company_id:
    company = self.company_id
strategy = company.country_restriction_strategy
```

File: product_country_restriction/models/product_country_restriction.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    )
+    resulting_country_ids = fields.Many2many(
+        comodel_name="res.country",
+        compute="_compute_resulting_country_ids",
+        readonly=True,
+    )
+
+    @api.model
+    def _default_company_id(self):
+        return self.env["res.company"]._company_default_get()
```
Comment (acsonefho) [inline product_country_restriction/models/product_country_restriction.py]: The `_company_default_get(...)` is deprecated.

File: product_country_restriction/models/product_country_restriction.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    @api.model
+    def _get_country_restriction_messages(self, restrictions_by_product):
+        """
+        Parse the applied embargo items per product and generate the
+        message
+        :param: {product_id: [(country, item)]}
+        :return:
+        """
+        messages = []
+        strategy = self.env.user.company_id.country_restriction_strategy
```
Comment (acsonefho) [inline product_country_restriction/models/product_country_restriction.py]: `self.env.company.country_restriction_strategy`

File: product_country_restriction/models/product_country_restriction_item.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                        fields.Date.to_string(rec.start_date),
+                        fields.Date.to_string(rec.end_date),
+                    ]
+                )
+            params = [rec.restriction_id.name, rec.rule_id.name]
+            if dates:
+                params.append(dates)
+            name = " - ".join(params)
+            rec.name = name
+        return res
```
Comment (acsonefho) [inline product_country_restriction/models/product_country_restriction_item.py]: A compute shouldn't `return` anything

File: product_country_restriction/readme/HISTORY.rst [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+* Allows to define different strategies by company:
+* * Authorize (default): Authorize all products except those with a
+    rule
+* * Restrict: Restrict all products except those with a rule
+
+10.0.1.1.1 (2019-01-30)
+~~~~~~~~~~~~~~~~~~~~~~~
+
+* The rule on product template must apply to its variants
```
Comment (acsonefho) [inline product_country_restriction/readme/HISTORY.rst:17]: Maybe you can add also your migration

File: product_country_restriction/tests/common.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class CountryRestrictionCommon(common.TransactionCase):
+    def setUp(self):
+        super(CountryRestrictionCommon, self).setUp()
```
Comment (acsonefho) [inline product_country_restriction/tests/common.py]: Python 3 syntax: `super().setUp()`

**Review by acsonefho — CHANGES_REQUESTED** [review]

**Review by acsonefho — APPROVED** [review]

**Review by FrancoMaxime — APPROVED** [review]

**Comment by xavier-bouquiaux** [comment]
@rousseldenis Can we merge this ?

**Pull request merged**